app/gui2/src/composables/nodeCreation.ts

```diff
@@ -127,29 +140,63 @@ export function useNodeCreation(
     onCreated(created)
   }
 
+  /** A callback that will be called with the created assignment expression.
+   * We resolve import conflicts and substitute identifiers if needed. */
+  function afterCreation(
+    edit: Ast.MutableModule,
+    ident: Ast.Identifier,
+    options: NodeCreationOptions,
+    identifiersRenameMap: Map<Ast.Identifier, Ast.Identifier>,
+  ) {
+    return (assignment: Ast.Assignment) => {
+      // When nodes are copied, we need to substitute original names with newly assigned.
+      if (options.binding) {
+        if (isIdentifier(options.binding) && options.binding !== ident)
+          identifiersRenameMap.set(options.binding, ident)
+      }
+      for (const [old, replacement] of identifiersRenameMap.entries()) {
+        substituteIdentifier(edit, assignment, old, replacement)
+      }
+
+      // Resolve import conflicts.
+      const conflicts = graphStore.addMissingImports(edit, options.requiredImports ?? []) ?? []
+      for (const _conflict of conflicts) {
+        // TODO: Substitution does not work, because we interpret imports wrongly. To be fixed in
+        // https://github.com/enso-org/enso/issues/9356
+        // substituteQualifiedName(edit, assignment, conflict.pattern, conflict.fullyQualified)
+      }
+    }
+  }
+
   function createNode(options: NodeCreationOptions) {
     createNodes([options])
   }
 
-  function newAssignmentNode(edit: Ast.MutableModule, options: NodeCreationOptions) {
-    const conflicts = graphStore.addMissingImports(edit, options.requiredImports ?? []) ?? []
-    const rhs = Ast.parse(options.expression, edit)
-    const inferredPrefix = inferPrefixFromAst(rhs)
-    const namePrefix = options.type ? typeToPrefix(options.type) : inferredPrefix
-    const ident = graphStore.generateLocallyUniqueIdent(namePrefix)
+  function newAssignmentNode(
+    edit: Ast.MutableModule,
+    ident: Ast.Identifier,
+    rhs: Ast.Owned,
+    fixup: (assignment: Ast.Assignment) => void,
+    options: NodeCreationOptions,
+  ) {
     rhs.setNodeMetadata(options.metadata ?? {})
     const assignment = Ast.Assignment.new(edit, ident, rhs)
-    for (const _conflict of conflicts) {
-      // TODO: Substitution does not work, because we interpret imports wrongly. To be fixed in
-      // https://github.com/enso-org/enso/issues/9356
-      // substituteQualifiedName(edit, assignment, conflict.pattern, conflict.fullyQualified)
-    }
+    fixup(assignment)
```

Comment:
I don't see the need for the complexity of injecting this logic as a callback. Why not call `afterCreation` directly in `createNodes`?

Reply:
My initial idea was to reduce the number of arguments, but it seems it won’t work, too much coupling in the code. Applied this suggestion.